src/api-service/__app__/onefuzzlib/webhooks.py

```diff
@@ -0,0 +1,222 @@
+import datetime
+import hmac
+import logging
+from hashlib import sha512
+from typing import List, Optional, Tuple
+from uuid import UUID
+
+import requests
+from memoization import cached
+from onefuzztypes.enums import ErrorCode, WebhookEventType, WebhookMessageState
+from onefuzztypes.models import Error, Result
+from onefuzztypes.webhooks import Webhook as BASE_WEBHOOK
+from onefuzztypes.webhooks import WebhookEvent, WebhookEventPing, WebhookMessage
+from onefuzztypes.webhooks import WebhookMessageLog as BASE_WEBHOOK_MESSAGE_LOG
+from pydantic import BaseModel
+
+from .__version__ import __version__
+from .azure.creds import get_func_storage
+from .azure.queue import queue_object
+from .orm import ORMMixin
+
+MAX_TRIES = 5
+EXPIRE_DAYS = 7
+USER_AGENT = "onefuzz-webhook %s" % (__version__)
+
+
+class WebhookMessageQueueObj(BaseModel):
+    webhook_id: UUID
+    event_id: UUID
+
+
+class WebhookMessageLog(BASE_WEBHOOK_MESSAGE_LOG, ORMMixin):
+    @classmethod
+    def key_fields(cls) -> Tuple[str, Optional[str]]:
+        return ("webhook_id", "event_id")
+
+    @classmethod
+    def search_expired(cls) -> List["WebhookMessageLog"]:
```

Comment:
Any thought on using a an azure queue instead of a table to store the message. we can take advantage of the message visibility for the retrying logic and the message expiration for and removing expired messages.

Reply:
There is a queue used for the actual orchestration.  It uses the message visibility to delay the retry.

However, by using a table for storing the underlying message log, we enable getting a list of recent deliveries.

Inspiration; [Github Webhook Events - Recent Deliveries](https://docs.github.com/en/free-pro-team@latest/developers/webhooks-and-events/testing-webhooks#listing-recent-deliveries)